lighthouse-core/audits/robust-csp.js

```diff
@@ -0,0 +1,188 @@
+  description: 'A strong Content Security Policy (CSP) can significantly ' +
+    'reduce the risk of XSS attacks. ' +
+    '[Learn more](https://developers.google.com/web/fundamentals/security/csp)',
+  /** Summary text for the results of a Lighthouse audit that evaluates the security of a page's CSP. This is displayed if no CSP is being enforced. "CSP" stands for "Content Security Policy". "CSP" does not need to be translated. */
+  noCsp: 'No CSP found in enforcement mode',
+  /** Warning message for a Lighthouse audit that evaluates the security of a page's CSP. This is displayed if no security vulnerabilities are found, but additional suggestions are available for tuning the CSP. "CSP" stands for "Content Security Policy". "CSP" does not need to be translated. */
+  additionalWarning: 'Additional CSP suggestions are available.',
+  /** Message shown when one or more CSPs are defined in a <meta> tag. Shown in a table with a list of other CSP vulnerabilities and suggestions. "CSP" stands for "Content Security Policy". "CSP" and "HTTP" do not need to be translated. */
+  metaTagMessage: 'The page contains a CSP defined in a <meta> tag. ' +
+    'It is not recommended to use a CSP this way, ' +
+    'consider defining the CSP in an HTTP header.',
+  /**
+   * @description [ICU Syntax] Message identifying a CSP which contains one or more syntax errors. Shown in a table with a list of other CSP vulnerabilities and suggestions. "CSP" stands for "Content Security Policy". "CSP" does not need to be translated.
+   * @example {script-src 'none'; object-src 'self';} rawCsp
+   */
+  syntaxMessage: `{numSyntax, plural,
+    =1 {Syntax error in CSP "{rawCsp}"}
+    other {Syntax errors in CSP "{rawCsp}"}
+  }`,
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class RobustCSP extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'robust-csp',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['devtoolsLogs', 'MetaElements', 'URL'],
+    };
+  }
+
+  /**
+   * @param {Finding} finding
+   * @return {LH.Audit.Details.TableItem}
+   */
+  static findingToTableItem(finding) {
+    return {
+      directive: finding.directive,
+      description: getTranslatedDescription(finding),
+    };
+  }
+
+  /**
+   * @param {Array<string>} rawCsps
+   * @return {LH.Audit.Details.TableItem[]}
+   */
+  static collectSyntaxResults(rawCsps) {
+    /** @type {LH.Audit.Details.TableItem[]} */
+    const results = [];
+
+    const syntaxFindingsByCsp = evaluateRawCspForSyntax(rawCsps);
+    for (let i = 0; i < rawCsps.length; ++i) {
+      const items = syntaxFindingsByCsp[i].map(this.findingToTableItem);
+      if (!items.length) continue;
```

Comment:
I think it could be useful to show a row for each `rawCsp`, even if there is no syntax error.

Reply:
I like that idea.

Maybe we get rid of the "Syntax errors in ..." part of it and just make the CSP the entire content of the row. Then each subitem would be "Syntax: Missing semicolon" or whatever.

Reply:
Quick preview of what this looks like:
<img width="907" alt="Screen Shot 2021-02-04 at 2 37 01 PM" src="https://user-images.githubusercontent.com/6752989/106946090-a6b80d80-66f6-11eb-95b7-9cea0cbcaf5e.png">

I realized we can get monospace by using "code" instead of "text".

Reply:
perhaps they should be first, and inside a single item as a subitem per CSP?

Reply:
Where would syntax errors go then? I think it's important to display syntax errors under the respective CSP.

Reply:
Honestly, I'm thinking we put a "No syntax errors found" under clean CSPs. It feels a little weird having the CSP sitting there with no other information. Maybe this wouldn't be a problem with severity icons though.

Reply:
This is where I got, moved directive column to the other side as well.
<img width="907" alt="Screen Shot 2021-02-04 at 4 16 08 PM" src="https://user-images.githubusercontent.com/6752989/106956225-52b42580-6704-11eb-9da9-04eec19f0058.png">